### PR TITLE
Add FnRow v1 keyboard by dinofizz

### PR DESCRIPTION
**Pull request by dinofizz**
## Description

Addition of a new keyboard - the FnRow v1, which features a single row of switches in the form of a “function row” layout.

All hardware design files are available under a CC-BY-SA license on GitHub [here](https://github.com/dinofizz/fnrow-pcb)

Blog post [here](https://www.dinofizzotti.com/blog/2021-02-07-fnrow-v1-a-configurable-function-row-layout-mechanical-keyboard/)

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/dinofizz/fnrow/v1/keymaps/via/keymap.c [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  //user initialization
+}
+
+void matrix_scan_user(void) {
+  //user matrix
+}
+
+bool process_record_user(uint16_t keycode, keyrecord_t *record) {
+  return true;
+}
```
Comment (drashna) [inline keyboards/dinofizz/fnrow/v1/keymaps/via/keymap.c:46]:
```suggestion
```

File: keyboards/dinofizz/fnrow/v1/readme.md [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+##  Features
+
+* USB-C
+* USB ESD and overcurrent protection
+* Dedicated reset switch to enable DFU mode
+* Optional status LED
+
+## Compilation and Flashing
+
+Make compilation example for this keyboard (after setting up your build environment):
```
Comment (drashna) [inline keyboards/dinofizz/fnrow/v1/readme.md:18]: Could you add instructions on how to enter bootloader mode?

File: keyboards/dinofizz/fnrow/v1/v1.c [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+void keyboard_post_init_user(void) {
+    // Blink the LED so we know everything is running OK
+    // Finish with LED OFF
+    palClearPad(GPIOA, LED_PIN);
+    wait_ms(100);
+    palSetPad(GPIOA, LED_PIN);
+    wait_ms(100);
+    palClearPad(GPIOA, LED_PIN);
+}
```
Comment (drashna) [inline keyboards/dinofizz/fnrow/v1/v1.c:32]: https://docs.qmk.fm/#/internals_gpio_control


```suggestion
void keyboard_pre_init_user(void) {
    // Immediately set the LED pin as an output and set it ON
    setPinOutput(A15);
    writePinHigh(A15);
}

void keyboard_post_init_user(void) {
    // Blink the LED so we know everything is running OK
    // Finish with LED OFF
    writePinLow(A15);
    wait_ms(100);
    writePinHigh(A15);
    wait_ms(100);
    writePinLow(A15);
}
```

File: keyboards/dinofizz/fnrow/v1/v1.h [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+#include "quantum.h"
+
+#define LAYOUT_fn_row( \
+    K000,       K001, K002, K003, K004,    K005, K006, K007, K008,    K009, K010, K011, K012 \
+) { \
+    { K000,  K001,  K002,  K003,  K004,  K005,  K006, }, \
+    { K007,  K008,  K009,  K010,  K011,  K012,  KC_NO }, \
+}
+
+#define LED_PIN 15
```
Comment (drashna) [inline keyboards/dinofizz/fnrow/v1/v1.h:28]:
```suggestion
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/dinofizz/fnrow/v1/readme.md (hunk 2 above)
Comment (dinofizz) [reply keyboards/dinofizz/fnrow/v1/readme.md:18]: Certainly, thank you for prompting me on this.

**Review by drashna — APPROVED** [review]

File: keyboards/dinofizz/fnrow/v1/keymaps/default/config.h [hunk 5]
```diff
@@ -0,0 +1,19 @@
+/* Copyright 2021 Dino Fizzotti
```
Comment (zvecr) [inline keyboards/dinofizz/fnrow/v1/keymaps/default/config.h:1]: The file should be removed

File: keyboards/dinofizz/fnrow/v1/rules.mk [hunk 6]
```diff
@@ -0,0 +1,16 @@
+MCU = STM32F042
```
Comment (zvecr) [inline keyboards/dinofizz/fnrow/v1/rules.mk:1]:
```suggestion
# MCU name
MCU = STM32F042

```

File: keyboards/dinofizz/fnrow/v1/v1.h [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#pragma once
+
+#include <hal.h>
```
Comment (zvecr) [inline keyboards/dinofizz/fnrow/v1/v1.h:18]:
```suggestion
```

File: keyboards/dinofizz/fnrow/v1/info.json [hunk 8]
```diff
@@ -0,0 +1,79 @@
+{
+  "keyboard_name": "FnRow v1",
+  "maintainer": "dinofizz",
+  "url": "https://github.com/dinofizz/fnrow-pcb",
+  "width": 15,
+  "height": 1,
+  "layouts": {
+    "LAYOUT_fn_row": {
+      "key_count": 13,
```
Comment (zvecr) [inline keyboards/dinofizz/fnrow/v1/info.json:9]:
```suggestion
```

**Review by zvecr — CHANGES_REQUESTED** [review]

File: keyboards/dinofizz/fnrow/v1/v1.c [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+#include "v1.h"
+
+void keyboard_pre_init_user(void) {
+    // Immediately set the LED pin as an output and set it ON
+    setPinOutput(A15);
+    writePinHigh(A15);
+}
```
Comment (zvecr) [inline keyboards/dinofizz/fnrow/v1/v1.c:22]:
```suggestion
void keyboard_pre_init_kb(void) {
    // Immediately set the LED pin as an output and set it ON
    setPinOutput(A15);
    writePinHigh(A15);
    
    keyboard_pre_init_user();
}
```

File: keyboards/dinofizz/fnrow/v1/v1.c [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+void keyboard_post_init_user(void) {
+    // Blink the LED so we know everything is running OK
+    // Finish with LED OFF
+    writePinLow(A15);
+    wait_ms(100);
+    writePinHigh(A15);
+    wait_ms(100);
+    writePinLow(A15);
```
Comment (zvecr) [inline keyboards/dinofizz/fnrow/v1/v1.c:31]:
```suggestion
void keyboard_post_init_kb(void) {
    // Blink the LED so we know everything is running OK
    // Finish with LED OFF
    writePinLow(A15);
    wait_ms(100);
    writePinHigh(A15);
    wait_ms(100);
    writePinLow(A15);
    
    keyboard_post_init_user();
```

File: keyboards/dinofizz/fnrow/v1/config.h [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+#define MATRIX_COLS 7
+
+#define MATRIX_ROW_PINS { A0, A1 }
+#define MATRIX_COL_PINS {B0, B1, B3, B4, B5, B6, B7 }
+
+#define DIODE_DIRECTION COL2ROW
+
+#define DEBOUNCE 5
+
+#define BOOTMAGIC_KEY_SALT KC_F12
```
Comment (zvecr) [inline keyboards/dinofizz/fnrow/v1/config.h:38]: Not needed as using lite?
```suggestion
```

**Review by zvecr — CHANGES_REQUESTED** [review]

File: keyboards/dinofizz/fnrow/v1/config.h (hunk 11 above)
Comment (dinofizz) [reply keyboards/dinofizz/fnrow/v1/config.h:38]: Re thinking this - if all lite gets me is the ability to jump into the bootloader then it is not necessary as I have a dedicated hardware RESET button. I will remove.

File: keyboards/dinofizz/fnrow/v1/rules.mk [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+#
+BACKLIGHT_ENABLE = no
+RGBLIGHT_ENABLE = no
+BOOTMAGIC_ENABLE = no
+MOUSEKEY_ENABLE = no				# Mouse keys
+EXTRAKEY_ENABLE = yes				# Audio control and System control
+CONSOLE_ENABLE = no					# Console for debug
+COMMAND_ENABLE = no    				# Commands for debug and configuration
+NKRO_ENABLE = no					# USB Nkey Rollover
+
```
Comment (fauxpark) [inline keyboards/dinofizz/fnrow/v1/rules.mk:15]:
```suggestion
# Build Options
#   change yes to no to disable
#
BOOTMAGIC_ENABLE = no       # Virtual DIP switch configuration
MOUSEKEY_ENABLE = no        # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
NKRO_ENABLE = no            # USB Nkey Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
BLUETOOTH_ENABLE = no       # Enable Bluetooth
AUDIO_ENABLE = no           # Audio output
```

File: keyboards/dinofizz/fnrow/v1/mcuconf.h [hunk 13]
```diff
@@ -0,0 +1,187 @@
+/*
```
Comment (fauxpark) [inline keyboards/dinofizz/fnrow/v1/mcuconf.h:1]: Please run `qmk chibios-confmigrate` on these files.

File: keyboards/dinofizz/fnrow/v1/config.h [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+
+#pragma once
+
+#include "config_common.h"
+
+#define VENDOR_ID       0xD1F1
+#define PRODUCT_ID      0x0100
+#define DEVICE_VER      0x0001
+#define MANUFACTURER    Dino Fizzotti
+#define PRODUCT         Dino Fizzotti FnRow v1
```
Comment (fauxpark) [inline keyboards/dinofizz/fnrow/v1/config.h:26]:
```suggestion
#define PRODUCT         FnRow v1
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/dinofizz/fnrow/v1/config.h (hunk 14 above)
Comment (dinofizz) [reply keyboards/dinofizz/fnrow/v1/config.h]: Changes made as requested.

File: keyboards/dinofizz/fnrow/v1/mcuconf.h (hunk 13 above)
Comment (dinofizz) [reply keyboards/dinofizz/fnrow/v1/mcuconf.h:1]: I have updated the chconf and mcuconf files with the output of this command (this was new info for me)

File: keyboards/dinofizz/fnrow/v1/rules.mk (hunk 12 above)
Comment (dinofizz) [reply keyboards/dinofizz/fnrow/v1/rules.mk]: Updated as requested.

File: keyboards/dinofizz/fnrow/v1/rules.mk [hunk 15]
```diff
@@ -0,0 +1,22 @@
+# MCU name
+MCU = STM32F042
+
+# Build Options
```
Comment (fauxpark) [inline keyboards/dinofizz/fnrow/v1/rules.mk:4]:
```suggestion
# Bootloader selection
BOOTLOADER = stm32-dfu

# Build Options
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/dinofizz/fnrow/v1/mcuconf.h (hunk 13 above)
Comment (fauxpark) [reply keyboards/dinofizz/fnrow/v1/mcuconf.h:1]: This seems to be still outstanding.

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/dinofizz/fnrow/v1/mcuconf.h (hunk 13 above)
Comment (dinofizz) [reply keyboards/dinofizz/fnrow/v1/mcuconf.h]: Thank you, apologies for wasting yoru time. I had updated chconf and halconf, not mcuconf. I have pushed a commit with the change. Are there any other conf files I should generate using this tool?

**Review by fauxpark — APPROVED** [review]

**Comment by drashna** [comment]
Thanks!

**Pull request merged**